**Question 1:  What is K-Nearest Neighbors (KNN) and how does it work in both
classification and regression problems?**


Ans =

**Question 1:  What is K-Nearest Neighbors (KNN) and how does it work in both
classification and regression problems?**

Ans = K-Nearest Neighbors (KNN) is a non-parametric, lazy learning algorithm that can be used for both classification and regression tasks.

**How it works:**

1.  **Training:** In the training phase, KNN simply stores the entire training dataset. There is no explicit model building.

2.  **Prediction:** To make a prediction for a new data point:
    *   Calculate the distance (usually Euclidean distance) between the new data point and all data points in the training set.
    *   Select the 'k' nearest neighbors (data points with the smallest distances).
    *   **For Classification:** The predicted class for the new data point is the class that is most frequent among the 'k' nearest neighbors.
    *   **For Regression:** The predicted value for the new data point is the average (or median) of the values of the 'k' nearest neighbors.

**Key aspects:**

*   **Parameter 'k':** The choice of 'k' is crucial. A small 'k' can be sensitive to noise, while a large 'k' can smooth out the decision boundary but may include data points from other classes/values.
*   **Distance Metric:** The distance metric used can influence the results (e.g., Euclidean, Manhattan).
*   **Computational Cost:** Prediction can be computationally expensive, especially for large datasets, as it requires calculating distances to all training points.
*   **No Explicit Model:** KNN is a "lazy" learner because it doesn't build an explicit model during training. The model is essentially the entire training dataset.

**Question 2: What is the Curse of Dimensionality and how does it affect KNN
performance?**

Ans =

**Question 2: What is the Curse of Dimensionality and how does it affect KNN
performance?**

Ans = The **Curse of Dimensionality** refers to the various difficulties that arise when working with data in high-dimensional spaces (i.e., data with a large number of features or attributes). As the number of dimensions increases, the volume of the space grows exponentially, and the available data becomes sparse within this vast space.

**How it affects KNN performance:**

*   **Increased Sparsity:** In high dimensions, data points become increasingly sparse. This means that the distance between any two points tends to be large, and the concept of "nearest neighbors" becomes less meaningful. All points can appear to be far away from each other.
*   **Distance Metrics Become Less Reliable:** In high dimensions, the difference in distance between the nearest and farthest neighbors can become smaller. This makes it harder to distinguish between neighbors and non-neighbors based on distance, leading to less reliable neighborhood identification.
*   **Increased Computational Cost:** Calculating distances between points in high dimensions is computationally more expensive.
*   **Overfitting:** With sparse data in high dimensions, KNN is more prone to overfitting. The model might find spurious patterns in the training data that don't generalize well to new, unseen data.
*   **Need for More Data:** To maintain the same density of data points in a high-dimensional space as in a lower-dimensional space, an exponentially larger amount of data is required. This is often impractical or impossible to obtain.

In summary, the Curse of Dimensionality makes it difficult for KNN to find true nearest neighbors, reduces the reliability of distance metrics, increases computational cost, and makes the model more susceptible to overfitting in high-dimensional spaces. This is why dimensionality reduction techniques are often used before applying KNN to high-dimensional data.

**Question 3: What is Principal Component Analysis (PCA)? How is it different from
feature selection?**

ans =

**Question 3: What is Principal Component Analysis (PCA)? How is it different from
feature selection?**

ans = **Principal Component Analysis (PCA)** is a dimensionality reduction technique that transforms a set of possibly correlated variables into a set of linearly uncorrelated variables called principal components. It does this by finding the directions (principal components) in the data that capture the most variance. The first principal component captures the most variance, the second captures the second most, and so on. By keeping only a subset of the principal components, we can reduce the dimensionality of the data while retaining as much of the original variance as possible.

**How it's different from feature selection:**

*   **PCA (Feature Extraction):** PCA creates new, synthetic features (principal components) that are linear combinations of the original features. The original features are transformed into a new set of features.
*   **Feature Selection:** Feature selection methods choose a subset of the original features to keep, discarding the rest. The features that are kept are the original features themselves.

In essence, PCA *extracts* new features, while feature selection *selects* existing features. PCA is useful when you want to reduce dimensionality while preserving as much information (variance) as possible, and the new, transformed features are acceptable. Feature selection is useful when you want to work with the original features, perhaps for interpretability or because the relationships between the original features are important.

**Question 4: What are eigenvalues and eigenvectors in PCA, and why are they
important?**



**Question 4: What are eigenvalues and eigenvectors in PCA, and why are they
important?**

Ans = In the context of PCA:

*   **Eigenvectors:** These are the directions or principal components along which the data varies the most. They are the directions of maximum variance in the data. In PCA, the eigenvectors of the covariance matrix of the data represent the principal components.
*   **Eigenvalues:** These are the magnitudes or "strengths" associated with each eigenvector. They indicate the amount of variance in the data that is captured by its corresponding eigenvector. A larger eigenvalue means that the corresponding eigenvector captures more of the data's variance.

**Why they are important in PCA:**

*   **Determining Principal Components:** Eigenvectors define the principal components. The eigenvector with the largest eigenvalue is the first principal component, the eigenvector with the second largest eigenvalue is the second principal component, and so on.
*   **Quantifying Variance:** Eigenvalues allow us to quantify how much variance is explained by each principal component. This is crucial for dimensionality reduction, as we can choose to keep only the principal components (eigenvectors) that correspond to the largest eigenvalues, thereby retaining the most important information (variance) in the data while reducing the number of dimensions.
*   **Ordering Principal Components:** Eigenvalues provide a way to order the principal components in terms of their importance (how much variance they capture). This ordering helps in deciding how many principal components to keep to achieve a desired level of dimensionality reduction and information retention.

In essence, eigenvalues and eigenvectors provide the mathematical foundation for PCA. They allow us to identify the most significant directions of variation in the data and quantify their importance, which is essential for dimensionality reduction while preserving as much information as possible.

****

**Question 5: How do KNN and PCA complement each other when applied in a single
pipeline?**

Ans =

**Question 5: How do KNN and PCA complement each other when applied in a single
pipeline?**

Ans = KNN and PCA complement each other effectively in a single pipeline, particularly when dealing with high-dimensional datasets. Here's how:

1.  **Addressing the Curse of Dimensionality:** PCA is a powerful technique for dimensionality reduction. By applying PCA before KNN, you can reduce the number of features in the dataset. This directly mitigates the negative effects of the Curse of Dimensionality on KNN performance, such as increased sparsity, less reliable distance metrics, and higher computational cost.
2.  **Improving KNN Performance:** Reducing dimensionality with PCA can lead to a more meaningful "nearest neighbor" search for KNN. In a lower-dimensional space where data is less sparse, the distances between points are more informative, allowing KNN to find true neighbors and make more accurate predictions.
3.  **Reducing Computational Load:** KNN's prediction phase involves calculating distances to all training points. In high dimensions, this is computationally expensive. By reducing the number of dimensions with PCA, the distance calculations become much faster, significantly reducing the overall computational cost of the KNN algorithm.
4.  **Noise Reduction:** PCA can also help in reducing noise in the data. By keeping only the principal components that capture the most variance, you are essentially filtering out the less important dimensions, which may contain noise. This can further improve the robustness of the KNN model.
5.  **Handling Multicollinearity:** If the original features are highly correlated, PCA can help by creating uncorrelated principal components. This can be beneficial for some distance metrics used in KNN.

In summary, PCA acts as a preprocessing step that prepares high-dimensional data for KNN. It addresses the challenges posed by the Curse of Dimensionality, leading to improved performance, reduced computational cost, and potentially better generalization for the KNN algorithm. The pipeline typically involves applying PCA to the data first and then training and using the KNN model on the transformed, lower-dimensional data.

**Dataset:
Use the Wine Dataset from sklearn.datasets.load_wine().
Question 6: Train a KNN Classifier on the Wine dataset with and without feature
scaling. Compare model accuracy in both cases.**

# Task
Train a KNN Classifier on the Wine dataset from `sklearn.datasets.load_wine()` with and without feature scaling and compare the model accuracy in both cases.

## Train knn without scaling

### Subtask:
Train a KNN classifier on the original data and evaluate its accuracy on the test set.


**Reasoning**:
The subtask requires loading the wine dataset, splitting it into training and testing sets, training a KNN classifier on the training data, making predictions on the test data, and calculating the accuracy. These steps can be grouped into a single code block.



In [2]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

# Load the wine dataset
wine = load_wine()
X = wine.data
y = wine.target

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize a KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)

# Train the KNN classifier on the training data
knn.fit(X_train, y_train)

# Make predictions on the test data
y_pred = knn.predict(X_test)

# Calculate the accuracy of the model
accuracy_original = accuracy_score(y_test, y_pred)

print(f"Accuracy without scaling: {accuracy_original}")

Accuracy without scaling: 0.7222222222222222


## Apply feature scaling

### Subtask:
Scale the features of the training and testing sets.


**Reasoning**:
Scale the training and testing data using StandardScaler as per the instructions.



In [3]:
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training data X_train
scaler.fit(X_train)

# Transform both X_train and X_test
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Data scaled successfully.")

Data scaled successfully.


## Train knn with scaling

### Subtask:
Train a KNN classifier on the scaled data and evaluate its accuracy on the scaled test set.


**Reasoning**:
Initialize a KNN classifier, train it on the scaled data, make predictions on the scaled test data, and calculate the accuracy.



In [4]:
# Initialize a new KNeighborsClassifier
knn_scaled = KNeighborsClassifier(n_neighbors=5)

# Train the KNN classifier on the scaled training data
knn_scaled.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_scaled = knn_scaled.predict(X_test_scaled)

# Calculate the accuracy of the model with scaling
accuracy_scaled = accuracy_score(y_test, y_pred_scaled)

# Print the accuracy of the model with scaling
print(f"Accuracy with scaling: {accuracy_scaled}")

Accuracy with scaling: 0.9444444444444444


## Compare the accuracies of the two models.

### Subtask:
Compare the accuracies of the two models.


**Reasoning**:
Print the calculated accuracies for both cases and provide a brief comparison.



In [5]:
print(f"Accuracy without scaling: {accuracy_original}")
print(f"Accuracy with scaling: {accuracy_scaled}")

if accuracy_scaled > accuracy_original:
    print("\nFeature scaling significantly improved the accuracy of the KNN model for this dataset.")
elif accuracy_scaled < accuracy_original:
    print("\nFeature scaling slightly decreased the accuracy of the KNN model for this dataset.")
else:
    print("\nFeature scaling did not change the accuracy of the KNN model for this dataset.")

Accuracy without scaling: 0.7222222222222222
Accuracy with scaling: 0.9444444444444444

Feature scaling significantly improved the accuracy of the KNN model for this dataset.


## Summary:

### Data Analysis Key Findings

*   The accuracy of the KNN model without feature scaling was found to be approximately 0.722.
*   The accuracy of the KNN model with feature scaling applied was approximately 0.944.

### Insights or Next Steps

*   Feature scaling significantly improved the accuracy of the KNN model for this dataset, increasing it from 72.2% to 94.4%. This highlights the importance of scaling for distance-based algorithms like KNN.
*   Further exploration could involve testing different scaling methods or optimizing the `n_neighbors` parameter for both the scaled and unscaled data.


**Question 7: Train a PCA model on the Wine dataset and print the explained variance
ratio of each principal component.**

In [6]:
from sklearn.decomposition import PCA
from sklearn.datasets import load_wine

# Load the wine dataset
wine = load_wine()
X = wine.data

# Apply PCA
pca = PCA()
pca.fit(X)

# Print the explained variance ratio of each principal component
print("Explained variance ratio of each principal component:")
for i, ratio in enumerate(pca.explained_variance_ratio_):
    print(f"Principal Component {i+1}: {ratio:.4f}")

Explained variance ratio of each principal component:
Principal Component 1: 0.9981
Principal Component 2: 0.0017
Principal Component 3: 0.0001
Principal Component 4: 0.0001
Principal Component 5: 0.0000
Principal Component 6: 0.0000
Principal Component 7: 0.0000
Principal Component 8: 0.0000
Principal Component 9: 0.0000
Principal Component 10: 0.0000
Principal Component 11: 0.0000
Principal Component 12: 0.0000
Principal Component 13: 0.0000


**Question 8: Train a KNN Classifier on the PCA-transformed dataset (retain top 2
components). Compare the accuracy with the original dataset.**

### Step 5: Compare accuracies

In [12]:
print(f"Accuracy with original data: {accuracy_original}")
print(f"Accuracy with PCA (2 components): {accuracy_pca}")

if accuracy_pca > accuracy_original:
    print("\nUsing PCA with 2 components improved the accuracy of the KNN model.")
elif accuracy_pca < accuracy_original:
    print("\nUsing PCA with 2 components slightly decreased the accuracy of the KNN model.")
else:
    print("\nUsing PCA with 2 components did not change the accuracy of the KNN model.")

Accuracy with original data: 0.7222222222222222
Accuracy with PCA (2 components): 0.7222222222222222

Using PCA with 2 components did not change the accuracy of the KNN model.


### Step 4: Evaluate KNN on PCA-transformed data

In [11]:
from sklearn.metrics import accuracy_score

# Make predictions on the PCA-transformed test data
y_pred_pca = knn_pca.predict(X_pca_test)

# Calculate the accuracy of the model on the PCA-transformed data
accuracy_pca = accuracy_score(y_test, y_pred_pca)

print(f"Accuracy with PCA (2 components): {accuracy_pca}")

Accuracy with PCA (2 components): 0.7222222222222222


### Step 3: Train KNN on PCA-transformed data

In [10]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize a KNeighborsClassifier
knn_pca = KNeighborsClassifier(n_neighbors=5)

# Train the KNN classifier on the PCA-transformed training data
knn_pca.fit(X_pca_train, y_train)

print("KNN classifier trained on PCA-transformed data.")

KNN classifier trained on PCA-transformed data.


### Step 2: Split PCA-transformed data

In [9]:
from sklearn.model_selection import train_test_split

# Split the PCA-transformed dataset into training and testing sets
X_pca_train, X_pca_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.2, random_state=42)

print("Shape of PCA training data:", X_pca_train.shape)
print("Shape of PCA testing data:", X_pca_test.shape)

Shape of PCA training data: (142, 2)
Shape of PCA testing data: (36, 2)


### Step 1: Apply PCA with 2 components

In [8]:
from sklearn.decomposition import PCA

# Apply PCA, retaining the top 2 principal components
pca_2 = PCA(n_components=2)
X_pca = pca_2.fit_transform(X)

print("Shape of original data:", X.shape)
print("Shape of PCA-transformed data:", X_pca.shape)

Shape of original data: (178, 13)
Shape of PCA-transformed data: (178, 2)


**Question 9: Train a KNN Classifier with different distance metrics (euclidean,
manhattan) on the scaled Wine dataset and compare the results.**

# Task
Train KNN classifiers with 'euclidean' and 'manhattan' distance metrics on the scaled Wine dataset, compare their accuracies, and summarize the findings.

## Train knn classifier (euclidean)

### Subtask:
Train a KNN classifier using the 'euclidean' distance metric on the scaled data.


**Reasoning**:
Initialize and train a KNN classifier with the euclidean distance metric on the scaled training data.



In [13]:
# Initialize a KNeighborsClassifier with euclidean distance
knn_euclidean = KNeighborsClassifier(n_neighbors=5, metric='euclidean')

# Train the KNN classifier on the scaled training data
knn_euclidean.fit(X_train_scaled, y_train)

print("KNN classifier with euclidean distance trained on scaled data.")

KNN classifier with euclidean distance trained on scaled data.


## Evaluate knn accuracy (euclidean)

### Subtask:
Evaluate the accuracy of the KNN model with the 'euclidean' distance metric.


**Reasoning**:
Evaluate the accuracy of the KNN model with the 'euclidean' distance metric on the scaled test set and print the result.



In [14]:
from sklearn.metrics import accuracy_score

# Make predictions on the scaled test data using the euclidean model
y_pred_euclidean = knn_euclidean.predict(X_test_scaled)

# Calculate the accuracy of the model with euclidean distance
accuracy_euclidean = accuracy_score(y_test, y_pred_euclidean)

# Print the accuracy
print(f"Accuracy with euclidean distance: {accuracy_euclidean}")

Accuracy with euclidean distance: 0.9444444444444444


## Train knn classifier (manhattan)

### Subtask:
Train a KNN classifier using the 'manhattan' distance metric on the scaled data.


**Reasoning**:
Initialize and train a KNN classifier with the 'manhattan' distance metric on the scaled training data.



In [15]:
# Initialize a KNeighborsClassifier with manhattan distance
knn_manhattan = KNeighborsClassifier(n_neighbors=5, metric='manhattan')

# Train the KNN classifier on the scaled training data
knn_manhattan.fit(X_train_scaled, y_train)

print("KNN classifier with manhattan distance trained on scaled data.")

KNN classifier with manhattan distance trained on scaled data.


## Evaluate knn accuracy (manhattan)

### Subtask:
Evaluate the accuracy of the KNN model with the 'manhattan' distance metric.


**Reasoning**:
Evaluate the accuracy of the KNN model with the 'manhattan' distance metric on the scaled test data and print the result.



In [16]:
# Make predictions on the scaled test data using the manhattan model
y_pred_manhattan = knn_manhattan.predict(X_test_scaled)

# Calculate the accuracy of the model with manhattan distance
accuracy_manhattan = accuracy_score(y_test, y_pred_manhattan)

# Print the accuracy
print(f"Accuracy with manhattan distance: {accuracy_manhattan}")

Accuracy with manhattan distance: 0.9444444444444444


## Compare model accuracies

### Subtask:
Compare the accuracies of the two models trained with different distance metrics.


**Reasoning**:
Print the accuracies obtained with euclidean and manhattan distance metrics and compare them.



In [17]:
print(f"Accuracy with euclidean distance: {accuracy_euclidean}")
print(f"Accuracy with manhattan distance: {accuracy_manhattan}")

if accuracy_euclidean > accuracy_manhattan:
    print("\nEuclidean distance resulted in higher accuracy.")
elif accuracy_manhattan > accuracy_euclidean:
    print("\nManhattan distance resulted in higher accuracy.")
else:
    print("\nBoth distance metrics resulted in the same accuracy.")

Accuracy with euclidean distance: 0.9444444444444444
Accuracy with manhattan distance: 0.9444444444444444

Both distance metrics resulted in the same accuracy.


## Summary:

### Data Analysis Key Findings

*   The KNN classifier trained with the 'euclidean' distance metric on the scaled data achieved an accuracy of approximately 0.9444.
*   The KNN classifier trained with the 'manhattan' distance metric on the scaled data also achieved an accuracy of approximately 0.9444.
*   Both distance metrics resulted in the same classification accuracy for the KNN model on this specific dataset.

### Insights or Next Steps

*   For this dataset, the choice between Euclidean and Manhattan distance metrics for KNN does not significantly impact accuracy. Further investigation could explore if this holds true with different `n_neighbors` values or other datasets.


**Question 10: You are working with a high-dimensional gene expression dataset to
classify patients with different types of cancer.
Due to the large number of features and a small number of samples, traditional models
overfit.
Explain how you would:
● Use PCA to reduce dimensionality
● Decide how many components to keep
● Use KNN for classification post-dimensionality reduction
● Evaluate the model
● Justify this pipeline to your stakeholders as a robust solution for real-world
biomedical data
(Include your Python code and output in the code box below.)**

## Addressing High-Dimensional Gene Expression Data with PCA and KNN

In the scenario of a high-dimensional gene expression dataset with a limited number of samples, where traditional models are prone to overfitting, a pipeline combining PCA for dimensionality reduction and KNN for classification is a robust approach. Here's how you would implement and justify this:

**1. Use PCA to reduce dimensionality:**

High-dimensional data often contains redundant or noisy features. PCA helps by transforming the data into a lower-dimensional space while preserving the most important variance. This is achieved by identifying the principal components, which are orthogonal directions in the feature space that capture the maximum variance.

**2. Decide how many components to keep:**

Selecting the optimal number of principal components is crucial. Several methods can be used:

*   **Explained Variance Ratio:** Plot the cumulative explained variance ratio as a function of the number of components. Choose the number of components that explain a significant portion of the total variance (e.g., 95%).
*   **Scree Plot:** Plot the eigenvalues (variance explained by each component) in descending order. Look for an "elbow" in the plot, where the rate of decrease in eigenvalues slows down. Components before the elbow are typically retained.
*   **Cross-validation:** Train and evaluate the KNN model with different numbers of principal components using cross-validation. Select the number of components that yields the best performance.

**3. Use KNN for classification post-dimensionality reduction:**

After reducing the dimensionality using PCA, you would train a KNN classifier on the transformed data. KNN is suitable here because:

*   It's a non-parametric algorithm, meaning it doesn't make assumptions about the underlying data distribution, which can be beneficial for complex biological data.
*   With reduced dimensionality, the "curse of dimensionality" is mitigated, and KNN can find meaningful nearest neighbors more effectively.

**4. Evaluate the model:**

Standard classification evaluation metrics should be used:

*   **Accuracy:** The proportion of correctly classified samples.
*   **Precision, Recall, F1-score:** These metrics are particularly important in biomedical applications where misclassifications can have significant consequences.
*   **Cross-validation:** Use k-fold cross-validation to get a more reliable estimate of the model's performance on unseen data and to avoid overfitting to the training set.

**5. Justify this pipeline to your stakeholders:**

This PCA+KNN pipeline is a robust solution for real-world biomedical data due to:

*   **Addressing Overfitting:** PCA effectively reduces the number of features, which is critical for preventing overfitting in scenarios with limited samples and high dimensionality.
*   **Improved Computational Efficiency:** Reducing dimensionality makes the KNN algorithm computationally less expensive, which is important for large datasets.
*   **Noise Reduction:** PCA can help filter out noise in the data by focusing on the components that capture the most variance.
*   **Interpretability (to some extent):** While the principal components themselves may not have direct biological interpretations, analyzing the loadings of the original features on the important components can provide insights into which genes contribute most to the variation in the data.
*   **Established Techniques:** Both PCA and KNN are well-established and widely used techniques in machine learning and data analysis, which can increase confidence in the results.

Here's a conceptual Python code example demonstrating this pipeline:

In [18]:
import numpy as np
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn.preprocessing import StandardScaler

# Conceptual Data (replace with your actual gene expression data)
# X: High-dimensional gene expression data (samples x features)
# y: Cancer type labels (samples)
# Assume X and y are loaded from your dataset

# For demonstration, let's create some dummy data
n_samples = 100
n_features = 1000
X = np.random.rand(n_samples, n_features)
y = np.random.randint(0, 3, n_samples) # 3 different cancer types

# 1. Scale the data (important for PCA)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 2. Use PCA to reduce dimensionality
# Decide number of components (e.g., based on explained variance)
pca = PCA(n_components=0.95) # Retain 95% of variance
X_pca = pca.fit_transform(X_scaled)

print(f"Original dimensions: {X.shape}")
print(f"Reduced dimensions after PCA: {X_pca.shape}")
print(f"Explained variance by selected components: {np.sum(pca.explained_variance_ratio_):.4f}")

# 3. Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.25, random_state=42)

# 4. Use KNN for classification post-dimensionality reduction
knn = KNeighborsClassifier(n_neighbors=5)

# Train the model
knn.fit(X_train, y_train)

# Make predictions
y_pred = knn.predict(X_test)

# 5. Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f"\nKNN Accuracy after PCA: {accuracy:.4f}")

# More detailed evaluation
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

# Cross-validation for robust evaluation
cv_scores = cross_val_score(knn, X_pca, y, cv=5) # 5-fold cross-validation
print(f"\nCross-validation Accuracy (5-fold): {np.mean(cv_scores):.4f} (+/- {np.std(cv_scores):.4f})")

# 6. Justification to stakeholders (covered in the markdown explanation)

Original dimensions: (100, 1000)
Reduced dimensions after PCA: (100, 90)
Explained variance by selected components: 0.9512

KNN Accuracy after PCA: 0.3200

Classification Report:
              precision    recall  f1-score   support

           0       0.20      0.14      0.17         7
           1       0.38      0.62      0.48         8
           2       0.29      0.20      0.24        10

    accuracy                           0.32        25
   macro avg       0.29      0.32      0.29        25
weighted avg       0.29      0.32      0.29        25


Cross-validation Accuracy (5-fold): 0.3200 (+/- 0.0748)
